projects/jdwp/codegen/dataclass_generator.py

```diff
@@ -59,9 +68,96 @@ def __generate_dataclass(self, struct: Struct) -> str:
             for field in struct.fields
             if self.__is_explicit_field(field)
         )
+
         class_def = f"@dataclasses.dataclass(frozen=True)\nclass {name}:\n{fields_def}"
+        class_def += "\n\n" + self.__generate_serialize_method(struct)
+        class_def += "\n\n" + self.__generate_parse_method(struct)
+
         return dedent(class_def)
 
+    def __generate_serialize_method(self, struct: Struct) -> str:
+        serialize_code = "def serialize(self, output: JDWPOutputStreamBase):\n"
+        for field in struct.fields:
+            serialize_code += f"    {self.__generate_serialize_field_code(field)}\n"
+        return dedent(serialize_code)
+
+    def __generate_serialize_array_code(self, field: Field) -> str:
+        array_length_field_type = field.type.length
+        array_code = (
+            f"output.write_{array_length_field_type.lower()}(len(self.{field.name}))\n"
+            f"for element in self.{field.name}:\n"
+            f"    element.serialize(output)\n"
+        )
+        return dedent(array_code)
+
+    def __generate_serialize_tagged_union_code(
+        self, field: Field, union: TaggedUnion
+    ) -> str:
+        union_code = "match self.{0}:\n".format(field.name)
+        for enum_val, struct_type in union.cases:
+            union_code += f"    case {struct_type.__name__}():\n"
+            union_code += f"        output.write_{union.tag.type.name.lower()}({enum_val.value})\n"
+            union_code += f"        self.{field.name}.serialize(output)\n"
+        return dedent(union_code)
+
+    def __generate_serialize_field_code(self, field: Field) -> str:
+        match field.type:
+            case Array():
+                return self.__generate_serialize_array_code(field)
+            case TaggedUnion():
+                return self.__generate_serialize_tagged_union_code(field)
+            case _:
+                return f"output.write_{field.type.name.lower()}(self.{field.name})"
+
+    def __generate_parse_method(self, struct: Struct) -> str:
+        struct_name = self.__struct_to_name[struct]
+        parse_code = f"@staticmethod\nasync def parse(input: JDWPInputStreamBase) -> {struct_name}:\n"
+
+        for field in struct.fields:
+            parse_code += (
+                f"    {field.name} = {self.__generate_parse_field_code(field)}\n"
```

Comment:
1. Try using `__get_field_name` here.
2. This does not work for arrays and unions, code generated for them by __generate_parse_field_code is not an expression.